### PR TITLE
feat(gyp): update gyp to v0.16.1

### PR DESCRIPTION
**Pull request by DeeDeeG**
## Checklist

- [x] `npm install && npm test` passes
- [ ] tests are included 
- [ ] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/googleapis/release-please#how-should-i-write-my-commits)
  - [ ] _(NOTE: The PR title does, which is enough if this is squash-and-merged. And otherwise, my own personal commits do. Some commits from before mine do not.)_

## Description of change

Bump `gyp-next` to v0.16.1.

The biggest reason for bumping to newer `gyp-next` at the moment is for **Python 3.12 support**, in my opinion.

## Refs

PRs:
Resolves: https://github.com/nodejs/node-gyp/pull/2912
Resolves: https://github.com/nodejs/node-gyp/pull/2909
Resolves: https://github.com/nodejs/node-gyp/pull/2825

Issues:
Resolves: https://github.com/nodejs/node-gyp/issues/2915
Resolves: https://github.com/nodejs/node-gyp/issues/2869

## Credit

Most of this work was done by others. I based this PR's branch off of @rzhao271's branch for the gyp-next v0.15.1 PR https://github.com/nodejs/node-gyp/pull/2909. See the gyp-next repo for the various contributors over there.

## Reviewers?

Pinging previously active folks around this gyp-next bump, who are probably waiting for this to PR be posted: @rzhao271, @cclauss, @lukekarrys

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -25,7 +25,7 @@ jobs:
       max-parallel: 15
       matrix:
         node: [16.x, 18.x, 20.x]
-        python: ["3.8", "3.11"]
+        python: ["3.8", "3.11", "3.12"]
```
Comment (cclauss) [inline .github/workflows/tests.yml:28]: Let's go with 8, 10, 12

**Review by rzhao271 — APPROVED** [review]

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -44,7 +44,7 @@ jobs:
       - name: Install Dependencies
         run: |
           npm install --no-progress
-          pip install pytest
+          python -m pip install pytest
```
Comment (cclauss) [inline .github/workflows/tests.yml:47]: Not needed in GHA

**Review by cclauss — APPROVED** [review]
A few minor bits but this looks good to me.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (cclauss) [inline .github/workflows/tests.yml:28]:
```suggestion
        python: ["3.8", "3.10", "3.12"]
```

File: .github/workflows/tests.yml (hunk 2 above)
Comment (cclauss) [inline .github/workflows/tests.yml:47]:
```suggestion
          pip install pytest
```

**Review by cclauss — APPROVED** [review]

File: .github/workflows/visual-studio.yml [hunk 3]
```diff
@@ -16,22 +16,22 @@ jobs:
...
-        msvs-version: [2016, 2019, 2022]  # https://github.com/actions/virtual-environments/tree/main/images/win
+        include:
+          - os: windows-2019
+            msvs-verison: 2019
+          - os: windows-2022
+            msvs-version: 2022
     runs-on: ${{ matrix.os }}
     steps:
       - name: Checkout Repository
         uses: actions/checkout@v3
```
Comment (cclauss) [inline .github/workflows/visual-studio.yml:27]:
```suggestion
        uses: actions/checkout@v4
```

**Review by lukekarrys — APPROVED** [review]

**Pull request merged**